src/main/java/io/aiven/kafka/connect/gcs/HourlyTopicPartitionRecordGrouper.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Aiven Kafka GCS Connector
+ * Copyright (c) 2019 Aiven Oy
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package io.aiven.kafka.connect.gcs;
+
+import java.time.LocalDateTime;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import io.aiven.kafka.connect.gcs.config.FilenameTemplateVariable;
+import io.aiven.kafka.connect.gcs.templating.Template;
+import io.aiven.kafka.connect.gcs.templating.Template.Instance;
+
+
+/**
+ * A {@link RecordGrouper} that groups records by topic and partition.
+ *
+ * <p>The class requires a filename template with {@code topic}, {@code partition},
+ * and {@code start_offset} variables declared.
+ *
+ * <p>The class supports limited and unlimited number of records in files.
+ */
+final class HourlyTopicPartitionRecordGrouper extends TopicPartitionRecordGrouper {
+    private static final List<String> EXPECTED_VARIABLE_LIST = Arrays.asList(
```

Comment:
Why not just construct the `HashSet` here?

Reply:
AFAIR you can't instantiate HashSet as a static variable, but you definitely can do so in the constructor - and it has been declared as well.

Creating one every call is wasteful either way.